collector/tasks.go

```diff
@@ -0,0 +1,174 @@
+// Copyright 2023 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package collector
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"path"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+type taskByAction struct {
+	Type   prometheus.ValueType
+	Desc   *prometheus.Desc
+	Value  func(action string, count int64) float64
+	Labels func(action string, count int64) []string
+}
+
+var (
+	taskLabels = []string{"cluster", "action"}
+)
+
+// Task Information Struct
+type Task struct {
+	logger  log.Logger
+	client  *http.Client
+	url     *url.URL
+	actions string
+
+	up                              prometheus.Gauge
+	totalScrapes, jsonParseFailures prometheus.Counter
+
+	byActionMetrics []*taskByAction
+}
+
+// NewTask defines Task Prometheus metrics
+func NewTask(logger log.Logger, client *http.Client, url *url.URL, actions string) *Task {
+	return &Task{
+		logger:  logger,
+		client:  client,
+		url:     url,
+		actions: actions,
+
+		up: prometheus.NewGauge(prometheus.GaugeOpts{
+			Name: prometheus.BuildFQName(namespace, "task_stats", "up"),
+			Help: "Was the last scrape of the ElasticSearch Task endpoint successful.",
+		}),
+		totalScrapes: prometheus.NewCounter(prometheus.CounterOpts{
+			Name: prometheus.BuildFQName(namespace, "task_stats", "total_scrapes"),
+			Help: "Current total Elasticsearch snapshots scrapes.",
+		}),
+		jsonParseFailures: prometheus.NewCounter(prometheus.CounterOpts{
+			Name: prometheus.BuildFQName(namespace, "task_stats", "json_parse_failures"),
+			Help: "Number of errors while parsing JSON.",
+		}),
+		byActionMetrics: []*taskByAction{
```

Comment:
I don't think we need a slice of metrics here. In fact, I think we can just declare the Desc as a package level var, and we can use it with `prometheus.MustNewConstMetric`. Having named vars for metrics helps with the linter finding unused metrics.

Reply:
Done